### PR TITLE
[COST-5098] - Fix usage counts and savings plan OCP on AWS

### PR DESCRIPTION
**Pull request by lcouzens**
## Jira Ticket

[COST-5098](https://issues.redhat.com/browse/COST-5098)

## Description

This change will address usage amount getting duplicated from Negation counts
This change also negates savings plan costs associated with OCP on AWS clusters.

## Testing Usage

1. Checkout Branch
2. Restart Koku
3. Create AWS provider with SavingsPlanCoveredUsage + SavingsPlanNegation line items. (I used some RHEL internal data via the customer filtered flow, reach out for help)
4. run the following queries: 

PG
`SELECT sum(usage_amount) FROM reporting_awscostentrylineitem_daily_summary;`

Trino
`select sum(lineitem_usageamount) from aws_line_items_daily;`

The difference between these totals should be equal to your SavingsPlanNegation usage 

`select sum(lineitem_usageamount) from aws_line_items_daily where lineitem_lineitemtype = 'SavingsPlanNegation';`

In my case the diff was 109

5. Run the same on main and see the usage has not negated the  SavingsPlanNegation

## Testing OCP on AWS cost
1. Checkout Branch
2. Restart Koku
3. Create OCP on AWS correlated data with SavingsPlanCoveredUsage line items
4. run following queries

Trino
'select sum(unblended_cost), resource_id from aws_openshift_daily_resource_matched_temp group by resource_id;'
vs
'select sum(lineitem_unblendedcost), lineitem_lineitemtype from aws_openshift_daily where lineitem_resourceid = 'filter in your resource_id' group by lineitem_lineitemtype;'

Cost should match usage only lineitems (negate any savings plan covered usage costs)

## Release Notes
- [ ] proposed release note

```markdown
* [COST-5098](https://issues.redhat.com/browse/COST-5098) Correctly attributes SavingsPlanCoveredUsage and SavingsPlanNegation line items to OCP on AWS tables.
* Examples below:

# Original Line Item Example:

| line_item_type           | usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost | resource_id
| --------                 | -------      | -------        | -------        | -------      | -------        | -------
| SavingsPlanCoveredUsage  | 1            | 2              | 2              | 2            | 2              | example_A
| SavingsPlanNegation      | 1            | -2             | -2             | -2           | -2             |

## AWS daily summary

During the AWS insert we set the usage amount to zero for the negation line to prevent duplicate usage counts. example:

| line_item_type           | usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost | resource_id
| --------                 | -------      | -------        | -------        | -------      | -------        | -------
| SavingsPlanCoveredUsage  | 1            | 2              | 2              | 2            | 2              | example_A
| SavingsPlanNegation      | 0            | -2             | -2             | -2           | -2             |

The aggregated result in the AWS table is as follows:

usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost 
-------      | -------        | -------        | -------      | ------- 
1            | 0             | 0            | 0           | 0

## AWS EC2
During the insert of the EC2 table we have the following filter:

`AND lineitem_resourceid != ''`

This filter excludes `SavingsPlanNegation` line items since they do not contain a resource_id. In order to correctly offset the negation to the relevant resource id we set the `SavingsPlanCoveredUsage` cost to zero: 

| line_item_type           | usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost | resource_id
| --------                 | -------      | -------        | -------        | -------      | -------        | -------
| SavingsPlanCoveredUsage  | 1            | 2              | 0              | 2            | 0              | example_A


## For OCP on AWS correlation
Again the `SavingsPlanNegation` rows do not have a resource_id. Therefore, they will not make it to the ocp on aws tables. So again we set the `SavingsPlanCoveredUsage` to zero in order to negate their cost.

| line_item_type           | usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost | resource_id
| --------                 | -------      | -------        | -------        | -------      | -------        | -------
| SavingsPlanCoveredUsage  | 1            | 2              | 0              | 2            | 0              | example_A




```

### CODE REVIEW

**Review by djnakabaale — APPROVED** [review]

**Review by djnakabaale — DISMISSED** [review]

**Comment by lcouzens** [comment]
Requires smoke test updates! QE are working on this effort.

File: koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql [hunk 1]
```diff
@@ -317,6 +323,13 @@ SELECT cast(uuid() as varchar) as uuid,
...
         END
         ) as unblended_cost,
+    sum(
+        CASE
+            WHEN aws.lineitem_lineitemtype='SavingsPlanCoveredUsage'
+            THEN 0.0
+            ELSE aws.lineitem_blendedcost
+        END
+        ) as blended_cost,
     sum(aws.lineitem_blendedcost) as blended_cost,
```
Comment (esebesto) [inline koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:333]: remove this line?

Comment (esebesto) [inline koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:333]: remove this line?

Comment (lcouzens) [reply koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:333]: Classic fail! YES! Thanks Eva :bow:

**Comment by myersCody** [comment]
If you are like me and looked at this pull request and asked yourself; "what are we doing here?". Maybe this will help:

# Original Line Item Example

| line_item_type           | usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost | resource_id
| --------                 | -------      | -------        | -------        | -------      | -------        | -------
| SavingsPlanCoveredUsage  | 1            | 2              | 2              | 2            | 2              | example_A
| SavingsPlanNegation      | 1            | -2             | -2             | -2           | -2             |

## AWS daily summary

During the AWS insert we set the usage amount to zero for the negation line:

| line_item_type           | usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost | resource_id
| --------                 | -------      | -------        | -------        | -------      | -------        | -------
| SavingsPlanCoveredUsage  | 1            | 2              | 2              | 2            | 2              | example_A
| SavingsPlanNegation      | 0            | -2             | -2             | -2           | -2             |

So the final result when they are aggregated is:
usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost 
-------      | -------        | -------        | -------      | ------- 
1            | 0             | 0            | 0           | 0

## AWS EC2
During the insert of the EC2 table we have the following filter:
```
AND lineitem_resourceid != ''
```
In theory, this filter should exclude the `SavingsPlanNegation` line items since they do not contain a resource_id. So now all we have to do is set the cost to zero: 

| line_item_type           | usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost | resource_id
| --------                 | -------      | -------        | -------        | -------      | -------        | -------
| SavingsPlanCoveredUsage  | 1            | 2              | 0              | 2            | 0              | example_A


## For OCP on AWS
Again the `SavingsPlanNegation` rows do not have a resource_id. Therefore, they will not make it to the ocp on cloud tables. Therefore, we just need to negate their cost.

| line_item_type           | usage_amount | unblended_rate | unblended_cost | blended_rate | unblended_cost | resource_id
| --------                 | -------      | -------        | -------        | -------      | -------        | -------
| SavingsPlanCoveredUsage  | 1            | 2              | 0              | 2            | 0              | example_A

**Review by djnakabaale — APPROVED** [review]

**Pull request merged**